### PR TITLE
Fix meta error in _convert_weight_to_int4pack

### PR DESCRIPTION
**Pull request by yanbing-j**
This PR is to fix meta error in _convert_weight_to_int4pack.

### CODE REVIEW

**Review by lezcano — COMMENTED** [review]
Is this not tested? Can this not tested in the relevant OpInfo? Can you add an OpInfo if the op doesn't have one?

**Comment by yanbing-j** [comment]
@lezcano This is not tested before. Since `_convert_weight_to_int4pack` is only for int4, using OpInfo is too heavy, I add a UT instead in `test_linaly.py`.

**Review by lezcano — COMMENTED** [review]
Sounds good, but I'll let @jerryzh168 review in case there is a more standard way to test this.

File: test/test_linalg.py [hunk 1]
```diff
@@ -6121,6 +6121,27 @@ def genf_int_float(x, y, use_transpose, non_contig_type):
...
+            self.skipTest("requires SM80 or later")
+
+        if TEST_WITH_ROCM:
+            if not CDNA2OrLater():
+                self.skipTest("_int4_mm is supported only for CDNA2 or later")
+
+        torch.manual_seed(1)
+        b = torch.rand((k, n), dtype=torch.bfloat16, device=device)
+        b_uint8, _ = _group_quantize_tensor(b, n_bit=4, q_group_size=32)
+        b_int4pack = torch._convert_weight_to_int4pack(b_uint8, innerKTiles=2)
```
Comment (jerryzh168) [inline test/test_linalg.py:6141]: nit: maybe parametrize the innerKTiles as well

**Review by jerryzh168 — APPROVED** [review]

File: test/test_linalg.py [hunk 2]
```diff
@@ -6121,6 +6121,27 @@ def genf_int_float(x, y, use_transpose, non_contig_type):
         torch._int_mm(a_int8, b_int8, out=c_int32_result)
         self.assertEqual(c_int32_result.float(), torch.mm(a_float, b_float))
 
+    @unittest.skipIf(IS_WINDOWS, "Skipped on Windows!")
+    @unittest.skipIf(IS_FBCODE and IS_REMOTE_GPU, "cublas runtime error")
+    @onlyNativeDeviceTypes
+    @parametrize("m", [32, 64])
+    @parametrize("k", [32, 64])
+    @parametrize("n", [48, 64])
+    def test__convert_weight_to_int4pack(self, device, m, k, n):
```
Comment (jerryzh168) [inline test/test_linalg.py:6130]: also why is this tested in test/test_linalg.py? is this the correct location?

**Comment by yanbing-j** [comment]
@lezcano @jerryzh168 
I try to add UT using OpInfo in `op_db`. `test_meta.py` can pass, unfortunately, `test_ops.py` cannot due to dtypes torch.int8, torch.int16, torch.int32, torch.int64. `_convert_weight_to_int4pack` only supports uint8 as input data type, but these dtypes will run in `test_ops.py` even if I add skip logic.
Could you please give some suggestions about this? Or is there any proper place to put the test? Thanks!

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -12156,6 +12162,36 @@ def sample_inputs_alias_copy(op_info, device, dtype, requires_grad, **kwargs):
                    'test_schema_correctness',
                    dtypes=(torch.complex64, torch.complex128)),
            )),
+    OpInfo('_convert_weight_to_int4pack',
+           dtypes=integral_types(),
```
Comment (lezcano) [inline torch/testing/_internal/common_methods_invocations.py:12166]: I think changing the following should fix the issue.
```suggestion
           dtypes=(torch.uint8,),
```

Comment (yanbing-j) [reply torch/testing/_internal/common_methods_invocations.py:12166]: Update to `dtypes=_dispatch_dtypes((torch.uint8,)),` instead.

**Comment by yanbing-j** [comment]
@lezcano CI still fails, but I cannot reproduce in A100. Do you have any idea? Thanks!
```
2024-07-20T07:01:33.3547919Z _ TestVmapOperatorsOpInfoCUDA.test_vmap_exhaustive__convert_weight_to_int4pack_cuda_uint8 _
2024-07-20T07:01:33.3548100Z Traceback (most recent call last):
2024-07-20T07:01:33.3549050Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_device_type.py", line 1127, in test_wrapper
2024-07-20T07:01:33.3549244Z     return test(*args, **kwargs)
2024-07-20T07:01:33.3550098Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_cuda.py", line 225, in wrapped
2024-07-20T07:01:33.3550258Z     return f(*args, **kwargs)
2024-07-20T07:01:33.3550879Z   File "/var/lib/jenkins/workspace/test/functorch/test_vmap.py", line 4242, in test_vmap_exhaustive
2024-07-20T07:01:33.3551047Z     self.opinfo_vmap_test(
2024-07-20T07:01:33.3551618Z   File "/var/lib/jenkins/workspace/test/functorch/test_vmap.py", line 4056, in opinfo_vmap_test
2024-07-20T07:01:33.3551770Z     test()
2024-07-20T07:01:33.3552246Z   File "/var/lib/jenkins/workspace/test/functorch/test_vmap.py", line 4034, in test
2024-07-20T07:01:33.3552424Z     self.vmap_outplace_test(
2024-07-20T07:01:33.3553006Z   File "/var/lib/jenkins/workspace/test/functorch/test_vmap.py", line 3942, in vmap_outplace_test
2024-07-20T07:01:33.3553314Z     for vmap_out, loop_out in compute_quantities_for_vmap_test(
2024-07-20T07:01:33.3554095Z   File "/var/lib/jenkins/workspace/test/functorch/common_utils.py", line 339, in compute_quantities_for_vmap_test
2024-07-20T07:01:33.3554372Z     for quantities in _compute_quantities_for_vmap_test(
2024-07-20T07:01:33.3555091Z   File "/var/lib/jenkins/workspace/test/functorch/common_utils.py", line 288, in _compute_quantities_for_vmap_test
2024-07-20T07:01:33.3555675Z     loop_out = loop(op, in_dims, out_dim, batch_size, *batched_args, **kwarg_values)
2024-07-20T07:01:33.3556164Z   File "/var/lib/jenkins/workspace/test/functorch/common_utils.py", line 42, in loop
2024-07-20T07:01:33.3556420Z     result = [torch.stack(out_lst) for out_lst in outs]
2024-07-20T07:01:33.3556977Z   File "/var/lib/jenkins/workspace/test/functorch/common_utils.py", line 42, in <listcomp>
2024-07-20T07:01:33.3557229Z     result = [torch.stack(out_lst) for out_lst in outs]
2024-07-20T07:01:33.3557487Z RuntimeError: CUDA error: invalid device function
2024-07-20T07:01:33.3558197Z CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
2024-07-20T07:01:33.3558466Z For debugging consider passing CUDA_LAUNCH_BLOCKING=1
2024-07-20T07:01:33.3559015Z Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.
```

**Comment by lezcano** [comment]
Alas, I don't work on this project anymore, so I won't be able to help landing this one. Probably @jerryzh168 can.

**Comment by yanbing-j** [comment]
@jerryzh168 Do you have any idea of the failure of `RuntimeError: CUDA error: invalid device function` mentioned above? Thanks!

File: aten/src/ATen/native/cuda/int4mm.cu [hunk 4]
```diff
@@ -1290,7 +1290,6 @@ at::Tensor _convert_weight_to_int4pack_cuda(
 
   TORCH_CHECK(in.dim() == 2);
   TORCH_CHECK(in.dtype() == at::kByte);
-  TORCH_CHECK(in.is_contiguous());
```
Comment (malfet) [inline aten/src/ATen/native/cuda/int4mm.cu:1293]: Why are you removing this check? I.e. it should only work for contiguous  tensors, shouldn't it?

Comment (yanbing-j) [reply aten/src/ATen/native/cuda/int4mm.cu:1293]: `python test/test_meta.py -k test_dispatch_symbolic_meta_outplace_all_strides__convert_weight_to_int4pack_` cannot pass this check. Can I remove this check and use `auto weight = in.contiguous()` instead?

Comment (malfet) [reply aten/src/ATen/native/cuda/int4mm.cu:1293]: May be right behavior would be to add `xfail` for all_strides tests, as this op is perf optimized and as such does not make much sense to use with non-contiguous tensors.

Comment (yanbing-j) [reply aten/src/ATen/native/cuda/int4mm.cu:1293]: Could you please tell me where to add `xfail` for all_strides tests?

Comment (yanbing-j) [reply aten/src/ATen/native/cuda/int4mm.cu:1293]: @malfet Could you please tell me where to add xfail for all_strides tests?

**Comment by yanbing-j** [comment]
Update this PR not to use OpInfo, to work for https://github.com/pytorch/ao/pull/517.

File: test/test_linalg.py [hunk 5]
```diff
@@ -6142,6 +6142,26 @@ def genf_int_float(x, y, use_transpose, non_contig_type):
         torch._int_mm(a_int8, b_int8, out=c_int32_result)
         self.assertEqual(c_int32_result.float(), torch.mm(a_float, b_float))
 
+    @unittest.skipIf(IS_WINDOWS, "Skipped on Windows!")
+    @unittest.skipIf(IS_FBCODE and IS_REMOTE_GPU, "cublas runtime error")
+    @onlyNativeDeviceTypes
+    def test__convert_weight_to_int4pack(self, device):
```
Comment (jerryzh168) [inline test/test_linalg.py:6148]: maybe you can link the issue in comment as well

**Review by jerryzh168 — APPROVED** [review]
LG, please make sure the CI passes before landing

File: test/test_linalg.py (hunk 5 above)
Comment (yanbing-j) [reply test/test_linalg.py:6148]: Done.

File: test/test_linalg.py [hunk 6]
```diff
@@ -6142,6 +6142,27 @@ def genf_int_float(x, y, use_transpose, non_contig_type):
...
+        # TODO: Fix https://github.com/pytorch/pytorch/issues/131425 and use OpInfo instead
+        test_list = [((64, 32), 2), ((64, 48), 2), ((64, 64), 2), ((256, 128), 4), ((256, 128), 8)]
+        if self.device_type == 'cuda' and not SM80OrLater:
+            self.skipTest("requires SM80 or later")
```
Comment (malfet) [inline test/test_linalg.py:6152]: Why this is the case?

**Pull request closed without merging**